tutorials/sphinx-tutorials/multiagent_ppo.py

```diff
@@ -0,0 +1,754 @@
+        n_agents=env.n_agents,
+        centralised=False,  # the policies are decentralized (ie each agent will act from its observation)
+        share_params=share_parameters_policy,
+        device=device,
+        depth=2,
+        num_cells=256,
+        activation_class=torch.nn.Tanh,
+    ),
+    NormalParamExtractor(),  # this will just separate the last dimension into two outputs: a loc and a non-negative scale
+)
+
+######################################################################
+# Wrap the neural network in a :class:`TensorDictModule`
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# This is simply a module that will read the ``in_keys`` it is provided with and write the
+# outputs of the neural network in-place at the ``out_keys``.
+#
+# Note that we use ``("agents", ...)`` keys as these keys are all per-agent.
+#
+policy_module = TensorDictModule(
+    policy_net,
+    in_keys=[("agents", "observation")],
+    out_keys=[("agents", "loc"), ("agents", "scale")],
+)
+
+######################################################################
+# Wrap the :class:`TensorDictModule` in a :class:`ProbabilisticActor`
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# We now need to build a distribution out of the location and scale of our
+# normal distribution. To do so, we instruct the :class:`ProbabilisticActor`
+# class to build a :class:`TanhNormal` out of the location and scale
+# parameters. We also provide the minimum and maximum values of this
+# distribution, which we gather from the environment specs.
+#
+# The name of the ``in_keys`` (and hence the name of the ``out_keys`` from
+# the :class:`TensorDictModule` above) has to end with the
+# :class:`TanhNormal` distribution constructor keyword arguments (loc and scale).
+#
+
+policy = ProbabilisticActor(
+    module=policy_module,
+    spec=env.unbatched_action_spec,
+    in_keys=[("agents", "loc"), ("agents", "scale")],
+    out_keys=[env.action_key],
+    distribution_class=TanhNormal,
+    distribution_kwargs={
+        "min": env.unbatched_action_spec[("agents", "action")].space.minimum,
+        "max": env.unbatched_action_spec[("agents", "action")].space.maximum,
+    },
+    return_log_prob=True,
+    log_prob_key=("agents", "sample_log_prob"),
+)  # we'll need the log-prob for the PPO loss
+
+######################################################################
+# Critic network
+# --------------
+#
+# The critic network is a crucial component of the PPO algorithm, even though it
+# isn't used at sampling time. This module will read the observations and
+# return the corresponding value.
+#
+# As before, you have to think carefully about the decision of **sharing the critic parameters**.
+# Sharing is definitely not recommended when agents have different reward functions. In cases
+# where the reward function (note: the reward function, NOT the reward) is the same for all agents (like here),
+# sharing might help performance.
```

Comment:
In general, even wit the same reward function it might have important impacts you need to consider